style_guides/angular_style_guide.md

```diff
@@ -63,3 +63,179 @@ app.controller('myController', function($scope, otherDeps, Private) {
 
 **Note:** Use this sparingly. Whenever possible, your modules should not be
 coupled to the angular application itself.
+
+## Directives
+
+### Use callbacks to communicate side effects up the hierarchy
+
+  - Treat attribute bindings as read-only and communicate with with the parent
+    directive via callbacks. Essentially treat directive attribute bindings
+    like function arguments, which should also not be mutated.
+
+    *Why?*: Spreading mutable state throughout the application makes it very
+    difficult to understand and predict. By receiving data from the parent
+    through the bindings and communicating desired changes back via separate
+    callback attributes the data flow becomes explicit. The attribute bindings
+    then become the public interface of the component, describing all the
+    inputs dictating the rendering and all possible effects on the application
+    state.
+
+  ```javascript
+  /* avoid */
+  module.directive('myBadTableHeader', function MyBadTableHeader() {
+    return {
+      bindToController: true,
+      controllerAs: 'tableHeader',
+      controller: TableHeaderController,
+      replace: true,
+      scope: {
+        caption: '=',
+        sortColumn: '=',
+      },
+      template: `<th
+        ng-class="{ 'tableHeader--sortColumn': tableHeader.isSortColumn() }"
+        ng-bind="tableHeader.caption"
+        ng-click="tableHeader.sort()"
+      ></th>`,
+    };
+  });
+
+  function TableHeaderController() {
+    this.isSortColumn = () => (
+      this.sortColumn === this.caption
+    );
+
+    this.sort = () => (
+      this.sortColumn = this.caption  // AVOID: mutating input attributes
+    );
+
+    /* OTHER STUFF */
+  }
+  ```
+
+  ```javascript
+  /* recommended */
+  module.directive('myGoodTableHeader', function MyGoodTableHeader() {
+    return {
+      bindToController: true,
+      controllerAs: 'tableHeader',
+      controller: TableHeaderController,
+      replace: true,
+      scope: {
+        caption: '=',          // INPUT
+        isSortColumn: '=',     // INPUT
+        onChangeSorting: '=',  // EFFECT
```

Comment:
I'm not necessarily advocating for this change, but what do you think about recommending expressions instead of binding a function as a value?

As far as I can tell, this would:
 - More clearly distinguish which attributes are functions inputs and which are effects
    For this directive the difference would be that the `on-change-sorting` attribute would look like this:
    ```html
    <th my-good-table-header on-change-sorting="controller.sortByColumnName(column.name)">
    ```
    Rather than:
    ```html
    <th my-good-table-header on-change-sorting="controller.sortByColumn">
    ```
- Allows the `myGoodTableHeader` directive to defines what values are available to the expression without having to determine a function signature that works for everyone
- Ensures that the `onChangeSorting` property bound to the controller is always a function, even if the consumer of `myGoodTableHeader` decides they don't want to use that capability
- May make documentation a little more necessary, as one can't just bind a function and use the debugger to see what arguments are passed but must walk back in the call stack to see what locals are injected